### ISSUE TITLE
Razer Naga keypad issue [BUG] 

### ISSUE THREAD

**Issue by zefrs** [RedBearAK/toshy#51]
Firstly, thanks so much for this app.  It is working a dream (besides the issue below :) )
```
Bare metal

DISTRO_NAME     = 'debian'
DISTRO_VER      = '12'
SESSION_TYPE    = 'x11'
DESKTOP_ENV     = 'gnome'

Any DE(s) tried with Wayland?: 

Keyboard type (IBM, Chromebook, Windows, Apple): 
Logitech USB Keyboard
Razer Razer Naga X
```

These terminal commands may provide helpful info:

- toshy-services-log
```
Showing systemd journal messages for Toshy services (since last boot):
Jul 27 18:14:12 PoopRoll systemd[2701]: Starting toshy-kde-dbus.service - Toshy KDE D-Bus Service...
Jul 27 18:14:15 PoopRoll systemd[2701]: toshy-kde-dbus.service: Control process exited, code=exited, status=1/FAILURE
Jul 27 18:14:15 PoopRoll systemd[2701]: toshy-kde-dbus.service: Failed with result 'exit-code'.
Jul 27 18:14:15 PoopRoll systemd[2701]: Failed to start toshy-kde-dbus.service - Toshy KDE D-Bus Service.
Jul 27 18:14:15 PoopRoll systemd[2701]: Starting toshy-config.service - Toshy Config Service...
Jul 27 18:14:15 PoopRoll systemd[2701]: Started toshy-session-monitor.service - Toshy Session Monitor.
Jul 27 18:14:15 PoopRoll systemd[2701]: Started toshy-config.service - Toshy Config Service.
Jul 27 18:14:21 PoopRoll systemd[2701]: toshy-kde-dbus.service: Scheduled restart job, restart counter is at 1.
Jul 27 18:14:21 PoopRoll systemd[2701]: Stopped toshy-kde-dbus.service - Toshy KDE D-Bus Service.
Jul 27 18:14:21 PoopRoll systemd[2701]: Starting toshy-kde-dbus.service - Toshy KDE D-Bus Service...
Jul 27 18:14:21 PoopRoll systemd[2701]: Started toshy-kde-dbus.service - Toshy KDE D-Bus Service.
Jul 27 20:34:22 PoopRoll systemd[2701]: Stopping toshy-session-monitor.service - Toshy Session Monitor...
Jul 27 20:34:22 PoopRoll systemd[2701]: Stopped toshy-session-monitor.service - Toshy Session Monitor.
Jul 27 20:34:22 PoopRoll systemd[2701]: toshy-session-monitor.service: Consumed 1min 18.298s CPU time.
Jul 27 20:34:23 PoopRoll systemd[2701]: Stopping toshy-config.service - Toshy Config Service...
Jul 27 20:34:23 PoopRoll bash[4042]: keyszer v0.6.92
Jul 27 20:34:23 PoopRoll bash[4042]: (--) WATCH: Watching for new devices to hot-plug.
Jul 27 20:34:23 PoopRoll bash[4042]: (--) Autodetecting all keyboards (--device not specified)
Jul 27 20:34:23 PoopRoll bash[4042]: (+K) Grabbing 'Razer Razer Naga X Keyboard' (/dev/input/event5)
Jul 27 20:34:23 PoopRoll bash[4042]: (+K) Grabbing 'Razer Razer Naga X' (/dev/input/event7)
Jul 27 20:34:23 PoopRoll bash[4042]: (+K) Grabbing 'Logitech USB Keyboard' (/dev/input/event8)
Jul 27 20:34:23 PoopRoll bash[4042]: (--) Ready to process input.
Jul 27 20:34:23 PoopRoll bash[4042]: signal TERM received
Jul 27 20:34:23 PoopRoll bash[4042]: (-K) Ungrabbing: 'Razer Razer Naga X Keyboard' (removed)
Jul 27 20:34:23 PoopRoll bash[4042]: (-K) Ungrabbing: 'Logitech USB Keyboard' (removed)
Jul 27 20:34:23 PoopRoll systemd[2701]: Stopped toshy-config.service - Toshy Config Service.
Jul 27 20:34:23 PoopRoll systemd[2701]: toshy-config.service: Consumed 41.010s CPU time.
```
Problem observed:  
* * *
When using my mouse number pad, it is registered as the keyboard strokes.  Below is the output, first key press is my #1 on keyboard, second is #1 on mouse.

```
(II) in KEY_1 (press)
(DD) on_key KEY_1 press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Logitech USB Keyboard'
(OO) press KEY_1 1690483297.996995
1
(II) in KEY_1 (release)
(DD) on_key KEY_1 release
(OO) release KEY_1 1690483298.0829184

(II) in KP1 (press)
(DD) modmap: KP1 => KEY_1 [Cond modmap - Forced Numpad feature]
(DD) on_key KEY_1 press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press KEY_1 1690483299.7756696
1
(II) in KP1 (release)
(DD) modmap: KP1 => KEY_1 [Cond modmap - Forced Numpad feature]
(DD) on_key KEY_1 release
(OO) release KEY_1 1690483299.869822
```

**Comment by RedBearAK** [comment]
@zefrs 

Interesting. A mouse with a bunch of numeric keys. But I notice it goes up to 12. I wonder how that works, because there's nothing beyond KP9 or KP0 in the key definition file. 

The "Forced Numpad" feature has to translate the keypad keys into "normal" numeric keys as a method of making the numpad behave like a numpad even if the NumLock LED is off. 

Try to disable the "Forced Numpad" feature from the tray icon menu (in the "Preferences" submenu) or using the GUI Toshy Preferences app if you don't have a working indicator tray. 

If you disable it that way, it should stay disabled, even if you reboot or restart Toshy. Or even if you reinstall Toshy, as it tries to save your feature settings. 

There is also a way to temporarily disable "Forced Numpad" when you want to use the navigation functions on a keypad, with Option+NumLock (or Fn+NumLock if you had a real Apple keyboard). Using that method it would revert back to being enabled if you restarted Toshy. This method would normally show a notification, but it looks like I broke that somehow and will need to fix it. 

Let me know if this doesn't solve your issue. There is also a modmap that deals with a bug in GTK3 apps involving keypad keys. That is only active when NumLock is OFF, but it may still affect you.

**Comment by RedBearAK** [comment]
@zefrs 

Also, it would be possible to have a modmap or keymap specific to that device, but it would be limited to remapping from the key press on the device to to some other key (or combo), or a macro sequence of keys and/or combos. If you have some software from Razer that works in Linux and deals with making the keys on the mouse do certain things, this is irrelevant. 

If you like having the "Forced Numpad" feature for the main keyboard, that modmap could be modified to be inactive only for key presses coming from the mouse device. This would let you keep it enabled for the main keyboard. 

If you were to find the modmap in the config and make it look like this, it should keep the mouse from being affected by that feature at all: 

```py
modmap("Cond modmap - Forced Numpad feature",{
    # Make numpad be a numpad regardless of Numlock state (like an Apple keyboard in macOS)
    Key.KP1:                    Key.KEY_1,
    Key.KP2:                    Key.KEY_2,
    Key.KP3:                    Key.KEY_3,
    Key.KP4:                    Key.KEY_4,
    Key.KP5:                    Key.KEY_5,
    Key.KP6:                    Key.KEY_6,
    Key.KP7:                    Key.KEY_7,
    Key.KP8:                    Key.KEY_8,
    Key.KP9:                    Key.KEY_9,
    Key.KP0:                    Key.KEY_0,
    Key.KPDOT:                  Key.DOT,  
    Key.KPENTER:                Key.ENTER,
}, when = lambda ctx: 
    matchProps(not_devn='Razer Razer Naga X') and
    matchProps(not_lst=remotes_lod)(ctx) and 
    cnfg.forced_numpad is True )
```

The difference is this line: 

```py
    matchProps(not_devn='Razer Razer Naga X') and
```

But this is a modification that you would need to restore from the backup the Toshy installer makes, if you ever reinstall or upgrade to a new version of Toshy. It wouldn't be automatically preserved and inserted into the new config file.

**Comment by zefrs** [comment]
Thanks so much for the feedback, i will give your suggestions a bash when i get home this evening.

**Comment by RedBearAK** [comment]
I was thinking about this, and if you can confirm that this exclusion solution works for you, I think it would be fine to include the fix in the default Toshy config file. That way you'd never need to keep track of the modification. 

I'll probably turn it into more of an exclusion list of device names, for similar devices with keypads that need to keep emitting keypad key codes. 

You will also probably need to test excluding your Razer device with the exact same condition from the modmap that does the GTK3 fix, to really get what you're looking for. 

This is probably what the whole thing would look like. Feel free to test this, while leaving the "Forced Numpad" feature enabled instead of disabled, to make sure the mouse gets ignored by these new modmap conditions: 

```py
# List of devices with keypads to exclude from Forced Numpad and GTK3 fix modmaps
exclude_kpad_devs_lod = [
    {devn:'Razer Razer Naga X'},
]


modmap("Cond modmap - Forced Numpad feature",{
    # Make numpad be a numpad regardless of Numlock state (like an Apple keyboard in macOS)
    Key.KP1:                    Key.KEY_1,
    Key.KP2:                    Key.KEY_2,
    Key.KP3:                    Key.KEY_3,
    Key.KP4:                    Key.KEY_4,
    Key.KP5:                    Key.KEY_5,
    Key.KP6:                    Key.KEY_6,
    Key.KP7:                    Key.KEY_7,
    Key.KP8:                    Key.KEY_8,
    Key.KP9:                    Key.KEY_9,
    Key.KP0:                    Key.KEY_0,
    Key.KPDOT:                  Key.DOT,  
    Key.KPENTER:                Key.ENTER,
}, when = lambda ctx: 
    matchProps(not_lst=exclude_kpad_devs_lod) and 
    matchProps(not_lst=remotes_lod)(ctx) and 
    cnfg.forced_numpad is True )


modmap("Cond modmap - GTK3 numpad nav keys fix",{
    # Make numpad nav keys work correctly in GTK3 apps
    # Key.KP5:                    Key.X,                          # GTK3 numpad fix - TEST TO SEE IF WORKING
    # Numpad PgUp/PgDn/Home/End keys
    Key.KP9:                    Key.PAGE_UP, 
    Key.KP3:                    Key.PAGE_DOWN, 
    Key.KP7:                    Key.HOME, 
    Key.KP1:                    Key.END,
    # Numpad arrow keys
    Key.KP8:                    Key.UP, 
    Key.KP2:                    Key.DOWN, 
    Key.KP4:                    Key.LEFT, 
    Key.KP6:                    Key.RIGHT,
    # Numpad Insert/Delete/Enter keys
    Key.KP0:                    Key.INSERT, 
    Key.KPDOT:                  Key.DELETE, 
    Key.KPENTER:                Key.ENTER,
# }, when = lambda ctx: ctx.numlock_on is False and forced_numpad is False)
# }, when = lambda ctx: ctx.numlock_on is False and cnfg.forced_numpad is False )
}, when = lambda ctx:
    matchProps(not_lst=exclude_kpad_devs_lod) and 
    matchProps(not_lst=remotes_lod)(ctx) and 
    matchProps(numlk=False)(ctx) and 
    cnfg.forced_numpad is False )
```

**Comment by zefrs** [comment]
Thanks @RedBearAK, just to confirm, do i just need to make these changes to toshy_config.py and restart the service?

**Comment by RedBearAK** [comment]
> do i just need to make these changes to toshy_config.py and restart the service?

Yep. And if it doesn't seem to come back, use the terminal command `toshy-config-verbose-start` to see a full debug output and see if you misplaced a comma or something. Python is pretty decent at actually pointing to which line has a problem. 

In my version of the Kinto config I've reworked some things to that all the options/preferences in the tray icon menu or the Preferences app are "live". But changing the contents of the config file to add a remap or disable something that isn't controlled by a variable will still require restarting the services. 

I tried to make it as simple as possible with the various terminal commands that do the same thing as the tray icon menu options.

**Comment by zefrs** [comment]
This does not seem to be working as expected.  I have added the config, however when pressing the mouse number pad i get the following:
```

(II) in KP2 (repeat)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat DOWN 1690652074.4527636

(II) in KP2 (repeat)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat DOWN 1690652074.4985058

(II) in KP2 (repeat)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat DOWN 1690652074.5394838

(II) in KP2 (repeat)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat DOWN 1690652074.5914514

(II) in KP2 (repeat)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat DOWN 1690652074.629009

(II) in KP2 (release)
(DD) modmap: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN release
(OO) release DOWN 1690652074.6319957

(II) in KP1 (press)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press END 1690652080.878353

(II) in KP1 (repeat)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat END 1690652081.1584263

(II) in KP1 (repeat)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat END 1690652081.1975892

(II) in KP1 (repeat)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat END 1690652081.2429705

(II) in KP1 (repeat)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat END 1690652081.2828813

(II) in KP1 (release)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END release
(OO) release END 1690652081.2845364

(II) in KP6 (press)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press RIGHT 1690652082.3756425

(II) in KP6 (release)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT release
(OO) release RIGHT 1690652082.584552

(II) in KP6 (press)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press RIGHT 1690652083.1083412

(II) in KP6 (repeat)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat RIGHT 1690652083.3672493

(II) in KP6 (repeat)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT repeat
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) repeat RIGHT 1690652083.4121442

(II) in KP6 (release)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT release
(OO) release RIGHT 1690652083.4133801

(II) in KP9 (press)
(DD) modmap: KP9 => PAGE_UP [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key PAGE_UP press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press PAGE_UP 1690652084.1359532

(II) in KP9 (release)
(DD) modmap: KP9 => PAGE_UP [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key PAGE_UP release
(OO) release PAGE_UP 1690652084.3579264

(II) in KP1 (press)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press END 1690652093.9297633

(II) in KP1 (release)
(DD) modmap: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END release
(OO) release END 1690652094.041579

(II) in KP5 (press)
(DD) on_key KP5 press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press KP5 1690652095.999822

(II) in KP5 (release)
(DD) on_key KP5 release
(OO) release KP5 1690652096.1358113

(II) in KP6 (press)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT press
(DD) KBTYPE: 'Windows' | Default type for dev (cached): 'Razer Razer Naga X'
(OO) press RIGHT 1690652097.373575

(II) in KP6 (release)
(DD) modmap: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT release
(OO) release RIGHT 1690652097.4710622
```

**Comment by RedBearAK** [comment]
🤦🏽 Sorry about that. Every 👏🏽  dang 👏🏽  time 👏🏽  I add a new function into the "when" conditional, where there are multiple conditions and a `lambda` wrapper function involved, I forget that the new function needs to be given the KeyContext object (ctx). This is not necessary if it isn't inside a lambda, so it always trips me up. 

Don't worry if none of that was comprehensible, just notice the small difference here between these two uses of `matchProps()`: 

```py
}, when = lambda ctx:
    matchProps(not_lst=exclude_kpad_devs_lod) and 
    matchProps(not_lst=remotes_lod)(ctx) and 
...
```

Just need to add the `(ctx)` so the new device exclusion list condition will actually do something. Without that it is literally operating on no information. So of course it isn't doing anything. 

```py
}, when = lambda ctx:
    matchProps(not_lst=exclude_kpad_devs_lod)(ctx) and 
    matchProps(not_lst=remotes_lod)(ctx) and 
...
```

Do this for both modmaps from the code block in the previous post. 

Since I made the same mistake on the same line in the first modmap conditional, I have to assume that you still have the "Forced Numpad" feature disabled. Make sure you test it with this change with the "Forced Numpad" feature enabled also. 

If this exclusion condition is working, you should not see the name of either of these modmaps showing up in the log when pressing the keypad keys on the mouse, and there should be no remapping of the KP* key codes anymore. 

Don't try to add the "(ctx)" to anything else like the line below, that is a direct comparison between a value in a variable and a boolean, so it doesn't need the context object. 

```py
    cnfg.forced_numpad is False )
```

Let me know if this still doesn't work.

**Comment by zefrs** [comment]
Apologies for the delay, this does seem to have resolved the issue.
Thanks so much for the assistance here.

**Comment by RedBearAK** [comment]
> Apologies for the delay, this does seem to have resolved the issue.

Glad it seems to be working. 

I hope that it will allow the keypad modmaps to continue working with the main keyboard, but just avoid operating on the mouse keypad. I will probably end up adding a few more devices like the Naga X to the default exclusion list over time. 

The change has actually been integrated into the default config file, so if you ever need to reinstall or just reinstall to get new features in the config, you shouldn't need to repeat the modification. 

I also put an editable "slice" in place so users can add a device on their own to be excluded from these modmaps. That should get protected and restored on each re-use of the Toshy installer. 

If you think this issue is satisfactorily resolved, feel free to close it yourself, so that you can easily re-open it if there seems to be a related issue. 

If a different issue crops up, put it in a new issue thread. 

>  Thanks so much for the assistance here.

Thanks for using Toshy and providing some feedback that improved it! 👍🏽

**Comment by zefrs** [comment]
Gonna bring up this issue again :)
Seems that i am having the same issue after reinstalling linux.  I have tried the above, but i think the condig vars have been updated since.

**Comment by zefrs** [comment]
@RedBearAK

**Comment by zefrs** [comment]
`II) in KP1 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END press
(DD) KBTYPE: 'Windows' | Default type for dev: 'Razer Razer Naga X'
(OO) press END 1740163044.809052
^[[F
(II) in KP1 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP1 => END [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key END release
(OO) release END 1740163044.965968

(II) in KP2 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN press
(DD) KBTYPE: 'Windows' | (CACHED) Default type for dev: 'Razer Razer Naga X'
(OO) press DOWN 1740163045.4552908
^[[B
(II) in KP2 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP2 => DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key DOWN release
(OO) release DOWN 1740163045.5599763

(II) in KP3 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP3 => PAGE_DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key PAGE_DOWN press
(DD) KBTYPE: 'Windows' | (CACHED) Default type for dev: 'Razer Razer Naga X'
(OO) press PAGE_DOWN 1740163045.9992406
^[[6~
(II) in KP3 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP3 => PAGE_DOWN [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key PAGE_DOWN release
(OO) release PAGE_DOWN 1740163046.138903

(II) in KP4 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP4 => LEFT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key LEFT press
(DD) KBTYPE: 'Windows' | (CACHED) Default type for dev: 'Razer Razer Naga X'
(OO) press LEFT 1740163047.893197
^[[D
(II) in KP4 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP4 => LEFT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key LEFT release
(OO) release LEFT 1740163048.0338523

(II) in KP5 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) on_key KP5 press
(DD) KBTYPE: 'Windows' | (CACHED) Default type for dev: 'Razer Razer Naga X'
(OO) press KP5 1740163048.4863806
5
(II) in KP5 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) on_key KP5 release
(OO) release KP5 1740163048.6157255

(II) in KP6 (press)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT press
(DD) KBTYPE: 'Windows' | (CACHED) Default type for dev: 'Razer Razer Naga X'
(OO) press RIGHT 1740163048.935195
^[[C
(II) in KP6 (release)
(CX) CTX_X11: Using Xlib for window context
(DD) MODMAP: KP6 => RIGHT [Cond modmap - GTK3 numpad nav keys fix]
(DD) on_key RIGHT release
(OO) release RIGHT 1740163049.0538566
`

**Comment by RedBearAK** [comment]
@zefrs 

Yes, the "lists of dictionaries" (`_lod`) turned out to be quite heavy performance-wise. So they've been restricted to being used only when the conditional expression needs to match on more than just a simple list of a single matching property. Everything else was reverted back to using a regex pattern string variable. And the simpler boolean variables and the regex pattern strings were placed before more complex matching conditions in all the `when` arguments, to short-circuit the condition as fast as possible. 

It was part of a number of different things I eventually did to drastically decrease the overall CPU usage of the default config file during typing. The results in the end were very different from the performance of the config from that long ago. "Night and day" would be a fitting phrase. 

Anyway, here's what the default config looks like now, in the section related to this issue: 

```py
###################################################################################################
###  SLICE_MARK_START: exclude_kpad_devs  ###  EDITS OUTSIDE THESE MARKS WILL BE LOST ON UPGRADE

# List of devices to add to the device exclusion list below this slice

exclude_kpad_devs_UserCustom_lst = [
    # Example syntax:
    # 'My Keyboard Device',

]

###  SLICE_MARK_END: exclude_kpad_devs  ###  EDITS OUTSIDE THESE MARKS WILL BE LOST ON UPGRADE
###################################################################################################


# DEPRECATED in favor of simple list:
# # List of devices with keypads to exclude from Forced Numpad and GTK3 fix modmaps
# exclude_kpad_devs_lod = [
#     {devn: 'Razer Razer Naga X'},
#     *exclude_kpad_devs_UserCustom_lod,
# ]

# Ensure the DEPRECATED variable exists and is a list before attempting to access it
# (New users WILL NOT have it in the marked slice above)
exclude_kpad_devs_UserCustom_lod = locals().get('exclude_kpad_devs_UserCustom_lod', [])

# Ensure the NEW variable exists and is a list before attempting to access it
# (Existing users MIGHT NOT have it in the marked slice above)
exclude_kpad_devs_UserCustom_lst = locals().get('exclude_kpad_devs_UserCustom_lod', [])

DEPRECATED_exclude_kpad_devs_UserCustom_lst = [
    device['devn'] for device in exclude_kpad_devs_UserCustom_lod 
    if isinstance(exclude_kpad_devs_UserCustom_lod, list)
]

# List of devices with keypads to exclude from Forced Numpad and GTK3 fix modmaps
exclude_kpad_devs_lst = [
    'Razer Razer Naga X',
    *DEPRECATED_exclude_kpad_devs_UserCustom_lst,
    *exclude_kpad_devs_UserCustom_lst,
]
exclude_kpad_devs_Str = toRgxStr(exclude_kpad_devs_lst)


modmap("Cond modmap - Forced Numpad feature",{
    # Make numpad be a numpad regardless of Numlock state (like an Apple keyboard in macOS)
    Key.KP1:                    Key.KEY_1,
    Key.KP2:                    Key.KEY_2,
    Key.KP3:                    Key.KEY_3,
    Key.KP4:                    Key.KEY_4,
    Key.KP5:                    Key.KEY_5,
    Key.KP6:                    Key.KEY_6,
    Key.KP7:                    Key.KEY_7,
    Key.KP8:                    Key.KEY_8,
    Key.KP9:                    Key.KEY_9,
    Key.KP0:                    Key.KEY_0,
    Key.KPDOT:                  Key.DOT,  
    Key.KPENTER:                Key.ENTER,
}, when = lambda ctx:
    cnfg.forced_numpad and
    matchProps(not_clas=exclude_kpad_devs_Str)(ctx) and
    matchProps(not_clas=remoteStr)(ctx)
)


modmap("Cond modmap - GTK3 numpad nav keys fix",{
    # Make numpad nav keys work correctly in GTK3 apps
    # Key.KP5:                    Key.X,                # GTK3 numpad fix - TEST TO SEE IF WORKING
    # Numpad PgUp/PgDn/Home/End keys
    Key.KP9:                    Key.PAGE_UP, 
    Key.KP3:                    Key.PAGE_DOWN, 
    Key.KP7:                    Key.HOME, 
    Key.KP1:                    Key.END,
    # Numpad arrow keys
    Key.KP8:                    Key.UP, 
    Key.KP2:                    Key.DOWN, 
    Key.KP4:                    Key.LEFT, 
    Key.KP6:                    Key.RIGHT,
    # Numpad Insert/Delete/Enter keys
    Key.KP0:                    Key.INSERT, 
    Key.KPDOT:                  Key.DELETE, 
    Key.KPENTER:                Key.ENTER,
}, when = lambda ctx:
    not cnfg.forced_numpad and
    matchProps(numlk=False)(ctx) and
    matchProps(not_clas=exclude_kpad_devs_Str)(ctx) and
    matchProps(not_clas=remoteStr)(ctx)
)
```

As you might see here, the 'Razer Razer Naga X' device name is still included in the newest default config file, outside of the 'exclude_kpad_devs' slice, and the editable slices purposely do not get updated by the Toshy installer. And the conditionals on the relevant modmaps are definitely different from that time period. 

```py
}, when = lambda ctx:
    cnfg.forced_numpad and
    matchProps(not_clas=exclude_kpad_devs_Str)(ctx) and
    matchProps(not_clas=remoteStr)(ctx)
)
```

```py
}, when = lambda ctx:
    not cnfg.forced_numpad and
    matchProps(numlk=False)(ctx) and
    matchProps(not_clas=exclude_kpad_devs_Str)(ctx) and
    matchProps(not_clas=remoteStr)(ctx)
)
```

You will probably be good if you remove the original structure you had inside the 'exclude_kpad_devs' and replace it with this to match the new default (without adding anything inside it):

```py
# List of devices to add to the device exclusion list below this slice

exclude_kpad_devs_UserCustom_lst = [
    # Example syntax:
    # 'My Keyboard Device',

]
```

And then either undo the edits you've made outside the slice (if you did any such edits), or reinstall Toshy once again, which would revert everything outside the editable slices to the new current state of the default config. (Which includes 'Razer Razer Naga X' automatically.)

Since I did make the attempt at backwards compatibility with the older "lod" style list inside the editable slice, I have to assume that you've made the edits to the modmaps outside that and ended up referencing the old 'remotes_lod' that was eventually removed, so reverting to the new default config via reinstalling should get you back on track. 

Let me know how it goes.

**Comment by RedBearAK** [comment]
The reason your edited config didn't just crash is because I haven't actually removed the deprecated "lod" lists yet, just stopped using them in the default config conditionals: 

```py
# DEPRECATED: Converted back to simple list format
# Add remote desktop clients & VM software here
remotes_lod = [
    {clas: "^Anydesk$"                       },
...
```

I will remove them at some point, when the number of users that might be inconvenienced by needing to start with a completely fresh default config file will be minimal.

**Comment by zefrs** [comment]
Sorry, i'm having a small brain day today... haha
Not exactly sure what i should do.
I have a fresh install of toshy, what changes should i make to the config to get it working?

**Comment by RedBearAK** [comment]
After a closer look using actual eyeballs I believe I identified a fundamental screw-up, and corrected it in the main branch. 

I was using the wrong argument name when trying to match on the device names. You can fix this yourself or grab a new zip file and reinstall. 

The two different modmap conditionals were using `not_clas=` on the `exclude_kpad_devs_Str` string, which is not a string generated from a list of app classes but rather a string generated from a list of device names. 

Make them both look like this, and it should work correctly. 

```py
}, when = lambda ctx:
    cnfg.forced_numpad and
    matchProps(not_devn=exclude_kpad_devs_Str)(ctx) and
    matchProps(not_clas=remoteStr)(ctx)
)
```

In other words, find the two lines that look like this:

```py
    matchProps(not_clas=exclude_kpad_devs_Str)(ctx) and
```

And change them to this: 

```py
    matchProps(not_devn=exclude_kpad_devs_Str)(ctx) and
```

This was my mistake all along. I didn't notice because I don't have the device for testing, and didn't have another similar device to use to check that the conditional expression was actually working correctly. I guess it's an obscure enough problem that nobody else ran into it and tried to report it. 

Yikes.

**Comment by zefrs** [comment]
BOOM!
That sorted it out, thanks so much!

**Comment by RedBearAK** [comment]
Thank goodness. 😅 

Closing again. 👍
